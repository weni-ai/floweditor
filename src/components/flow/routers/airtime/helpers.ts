import {
  AirtimeRouterFormState,
  AirtimeTransferEntry,
} from 'components/flow/routers/airtime/AirtimeRouterForm';
import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { TransferAirtime } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (
  settings: NodeEditorSettings,
): AirtimeRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName = { value: 'Result' };
  let valid = false;

  const amounts: AirtimeTransferEntry[] = [];
  if (originalAction && originalAction.type === Types.transfer_airtime) {
    Object.keys(originalAction.amounts).forEach((key: string) => {
      amounts.push({
        value: { code: key, amount: '' + originalAction.amounts[key] },
      });
    });
    resultName = { value: originalAction.result_name };
    valid = true;
  }

  return {
    valid,
    amounts,
    resultName,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: AirtimeRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const amounts = {};
  state.amounts.forEach((entry: AirtimeTransferEntry) => {
    if (entry.value.amount.trim().length > 0) {
      (amounts as any)[entry.value.code] = Number(entry.value.amount);
    }
  });

  const newAction: TransferAirtime = {
    uuid,
    type: Types.transfer_airtime,
    amounts,
    result_name: state.resultName.value,
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};

export const getOriginalAction = (
  settings: NodeEditorSettings,
): TransferAirtime => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.transfer_airtime) {
    return action as TransferAirtime;
  }
};
