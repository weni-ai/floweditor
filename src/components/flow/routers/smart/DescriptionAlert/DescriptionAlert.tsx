import * as React from 'react';

import styles from './DescriptionAlert.module.scss';

import { applyVueInReact } from 'vuereact-combined';

// @ts-ignore
import { unnnicIcon } from '@weni/unnnic-system';
import { Trans } from 'react-i18next';

const UnnnicIcon = applyVueInReact(unnnicIcon);

export const leadInSpecId = 'lead-in';

export interface DescriptionAlertProps {
  openDescriptionEdit: () => void;
}

export default class DescriptionAlert extends React.Component<DescriptionAlertProps> {
  public render(): JSX.Element {
    return (
      <div className={styles.alert} onClick={this.props.openDescriptionEdit}>
        <UnnnicIcon
          className={styles.icon}
          icon="info"
          size="sm"
          scheme="aux-blue-500"
          filled={true}
        />
        <span className={styles.alert_text}>
          <Trans i18nKey="forms.smart_wait.description_alert">
            <b>Important:</b>Edit your project, <u>add a description</u> and make your AI even more
            powerful.
          </Trans>
        </span>
      </div>
    );
  }
}
