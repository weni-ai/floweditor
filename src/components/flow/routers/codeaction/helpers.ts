import { createRenderNode } from 'components/flow/routers/helpers';
import {
  Category,
  Exit,
  RouterTypes,
  SwitchRouter,
  WebhookExitNames,
} from 'flowTypes';
import { Types } from 'config/interfaces';
import { RunCodeAction } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID, snakify } from 'utils';
import { CodeActionRouterFormState } from 'components/flow/routers/codeaction/CodeActionRouterForm';
import { StringEntry } from '../../../../store/nodeEditor';

export const getOriginalAction = (
  settings: NodeEditorSettings,
): RunCodeAction => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.run_code_action) {
    return action as RunCodeAction;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  initialCodeAction: any,
): CodeActionRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName: StringEntry = { value: 'Result' };
  let codeAction: FormEntry = {
    value: { id: 0, name: 'Select a Code Action' },
  };

  if (originalAction && originalAction.type === Types.run_code_action) {
    codeAction = { value: originalAction.codeaction };
    resultName = { value: originalAction.result_name };
  } else {
    if (initialCodeAction) {
      codeAction = initialCodeAction;
    }
  }

  const state: CodeActionRouterFormState = {
    resultName,
    codeAction,
    valid: true,
  };
  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: CodeActionRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }
  let categories: Category[] = [];
  let exits: Exit[] = [];

  const newAction: RunCodeAction = {
    uuid,
    type: Types.run_code_action,
    codeaction: {
      id: state.codeAction.value.id,
      name: state.codeAction.value.name,
    },
    result_name: state.resultName.value,
  };

  const originalNode = settings.originalNode;
  if (
    originalNode &&
    originalNode.node.actions.length === 1 &&
    originalNode.node.actions[0].type === newAction.type
  ) {
    exits = originalNode.node.exits;
    categories = originalNode.node.router.categories;
  } else {
    exits = [
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
    ];

    categories = [
      {
        uuid: createUUID(),
        name: WebhookExitNames.Success,
        exit_uuid: exits[0].uuid,
      },
      {
        uuid: createUUID(),
        name: WebhookExitNames.Failure,
        exit_uuid: exits[1].uuid,
      },
    ];
  }

  let operand = '@results.' + snakify(newAction.result_name);

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: categories[0].uuid,
    cases: [],
    categories,
    operand,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.run_code_action,
    [newAction],
  );

  return newRenderNode;
};
