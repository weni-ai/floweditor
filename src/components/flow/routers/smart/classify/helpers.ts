import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { AutomaticClassifyRouterFormState } from 'components/flow/routers/smart/classify/AutomaticClassifyRouterForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Type, Types, VISIBILITY_HIDDEN } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { ensureRoute } from '../../classify/helpers';
import { getOperatorConfig } from 'config';

export const nodeToState = (
  settings: NodeEditorSettings,
): AutomaticClassifyRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let operand = DEFAULT_OPERAND;

  let hiddenCases: CaseProps[] = [];

  if (
    settings.originalNode &&
    getType(settings.originalNode) === Types.automatic_classify
  ) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);

        hiddenCases = initialCases.filter(
          (kase: CaseProps) =>
            getOperatorConfig(kase.kase.type).visibility === VISIBILITY_HIDDEN,
        );

        initialCases = initialCases.filter(
          (kase: CaseProps) =>
            getOperatorConfig(kase.kase.type).visibility !== VISIBILITY_HIDDEN,
        );
      }

      resultName = { value: router.result_name || '' };
      operand = router.operand;
    }
  }

  return {
    cases: initialCases,
    hiddenCases,
    resultName,
    valid: true,
    operand: { value: operand },
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  typeConfig: Type,
  state: AutomaticClassifyRouterFormState,
): RenderNode => {
  const routes = resolveRoutes(
    [...state.cases, ...state.hiddenCases],
    false,
    settings.originalNode.node,
    'Failure',
  );

  // make sure we have an other route since failure is our default
  ensureRoute(routes, {
    type: Operators.has_category,
    arguments: [],
    name: 'Other',
  });

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const routerResultName = state.resultName.value;

  const router: SwitchRouter = {
    type: RouterTypes.smart,
    default_category_uuid: routes.defaultCategory,
    cases: routes.cases,
    categories: routes.categories,
    operand: state.operand.value,
    result_name: routerResultName,
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    routes.exits,
    typeConfig.type,
    [],
    { cases: routes.caseConfig },
  );

  return newRenderNode;
};
