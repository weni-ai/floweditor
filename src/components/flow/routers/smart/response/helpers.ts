import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { SmartResponseRouterFormState } from 'components/flow/routers/smart/response/SmartResponseRouterForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Type, Types, VISIBILITY_HIDDEN } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter, Wait, WaitTypes } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { ensureRoute } from '../../classify/helpers';
import { getOperatorConfig } from 'config';

export const nodeToState = (
  settings: NodeEditorSettings,
): SmartResponseRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let timeout = 0;

  let hiddenCases: CaseProps[] = [];

  if (
    (settings.originalNode &&
      getType(settings.originalNode) === Types.wait_for_response) ||
    getType(settings.originalNode) === Types.smart_wait_for_response
  ) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);

        hiddenCases = initialCases.filter(
          (kase: CaseProps) =>
            getOperatorConfig(kase.kase.type).visibility === VISIBILITY_HIDDEN,
        );

        initialCases = initialCases.filter(
          (kase: CaseProps) =>
            getOperatorConfig(kase.kase.type).visibility !== VISIBILITY_HIDDEN,
        );
      }

      resultName = { value: router.result_name || '' };
    }

    if (
      settings.originalNode.node.router.wait &&
      settings.originalNode.node.router.wait.timeout
    ) {
      timeout = settings.originalNode.node.router.wait.timeout.seconds || 0;
    }
  }

  return {
    cases: initialCases,
    hiddenCases,
    resultName,
    timeout,
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  typeConfig: Type,
  state: SmartResponseRouterFormState,
): RenderNode => {
  const hasFilledCases =
    state.cases.length > 0 && state.cases[0].categoryName.trim() !== '';

  const routes = resolveRoutes(
    [...state.cases, ...state.hiddenCases],
    state.timeout > 0,
    settings.originalNode.node,
    hasFilledCases ? 'Failure' : null,
  );

  if (hasFilledCases) {
    // make sure we have an other route since failure is our default
    ensureRoute(routes, {
      type: Operators.has_category,
      arguments: [],
      name: 'Other',
    });
  }

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.msg } as Wait;
  if (state.timeout > 0) {
    wait.timeout = {
      seconds: state.timeout,
      category_uuid: routes.timeoutCategory,
    };
  }

  const router: SwitchRouter = {
    type: RouterTypes.smart,
    default_category_uuid: routes.defaultCategory,
    cases: routes.cases,
    categories: routes.categories,
    operand: DEFAULT_OPERAND,
    wait,
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    routes.exits,
    typeConfig.type,
    [],
    { cases: routes.caseConfig },
  );

  return newRenderNode;
};
