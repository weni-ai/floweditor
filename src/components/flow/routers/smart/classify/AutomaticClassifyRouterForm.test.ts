import { RouterFormProps } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import AutomaticClassifyRouterFormState from 'components/flow/routers/smart/classify/AutomaticClassifyRouterForm';
import DescriptionAlert from 'components/flow/routers/smart/DescriptionAlert/DescriptionAlert';
import { Operators } from 'config/interfaces';
import { Types } from 'config/interfaces';
import { composeComponentTestUtils, mock } from 'testUtils';
import { getRouterFormProps, createMatchRouter } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { createUUID } from 'utils';
import { getSmartOrSwitchRouter } from 'components/flow/routers/helpers';
import { vi } from 'vitest';
import { shallowToJson } from 'enzyme-to-json';

const routerNode = createMatchRouter(['Red']);

const { setup } = composeComponentTestUtils<RouterFormProps>(
  AutomaticClassifyRouterFormState,
  getRouterFormProps(routerNode),
);

describe(AutomaticClassifyRouterFormState.name, () => {
  beforeEach(() => {
    mock(utils, 'createUUID', utils.seededUUIDs());
  });

  it('should render', () => {
    const { wrapper } = setup(true);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('initializes case config', () => {
    const baseCase = createUUID();

    const baseNode = createMatchRouter(['Red']);
    const router = getSmartOrSwitchRouter(baseNode.node);
    router.cases.push({
      uuid: baseCase,
      type: Operators.has_any_word,
      arguments: ['not blue'],
      category_uuid: router.categories[0].uuid,
    });

    baseNode.ui = {
      position: { left: 0, top: 0 },
      type: Types.automatic_classify,
      config: { cases: { [baseCase]: { arguments: ['not blue'] } } },
    };

    const { wrapper } = setup(true, {
      nodeSettings: {
        $set: {
          originalNode: baseNode,
        },
      },
    });

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });

      instance.handleOperandUpdated('@results.result1');
      instance.handleCasesUpdated([
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['red'] },
          categoryName: 'Red',
          valid: true,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['maroon'] },
          categoryName: 'Red',
          valid: true,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['green'] },
          categoryName: 'Green',
          valid: true,
        },
      ] as CaseProps[]);
      instance.handleUpdateResultName('Favorite Color');

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchSnapshot();
    });

    it('should not save changes if case is invalid', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });

      instance.handleOperandUpdated('@results.result1');
      instance.handleUpdateResultName('Favorite Color');
      instance.handleCasesUpdated([
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['red'] },
          categoryName: 'Red',
          valid: true,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['maroon'] },
          categoryName: 'Red',
          valid: false,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['green'] },
          categoryName: 'Green',
          valid: true,
        },
      ] as CaseProps[]);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).not.toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled;
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });
      instance.handleOperandUpdated('@results.result1');
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });

  it('should open description edit', () => {
    const { wrapper, instance } = setup(true);
    const spy = vi.spyOn(window.parent, 'postMessage');

    const descriptionAlert = wrapper.find(DescriptionAlert);
    expect(descriptionAlert.exists()).toBeTruthy();
    wrapper.find(DescriptionAlert).prop('openDescriptionEdit')();

    expect(spy).toHaveBeenCalledWith({ event: 'openConnectEditProject' }, '*');

    expect(instance.state.hasDescription).toBeFalsy();
    // send the message back
    window.dispatchEvent(
      new MessageEvent('message', {
        data: {
          event: 'setConnectProjectDescription',
          connectProjectDescription: 'test',
        },
      }),
    );
    expect(instance.state.hasDescription).toBe(true);
  });
});
