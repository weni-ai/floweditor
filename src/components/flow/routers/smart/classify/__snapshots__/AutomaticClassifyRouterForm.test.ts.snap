// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutomaticClassifyRouterForm > initializes case config 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="wait_for_response"
  title="Wait for response"
>
  <TypeList
    __className=""
    initialType={
      {
        "aliases": [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "categories",
          "cases",
        ],
        "name": "Wait for response",
        "type": "wait_for_response",
        "visibility": [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    nodeSettings={
      {
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [],
            "exits": [
              {
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
              {
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
            ],
            "router": {
              "cases": [
                {
                  "arguments": [
                    "red",
                  ],
                  "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "type": "has_any_word",
                  "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                },
                {
                  "arguments": [
                    "not blue",
                  ],
                  "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "type": "has_any_word",
                  "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
                },
              ],
              "categories": [
                {
                  "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                  "name": "Red",
                  "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                },
                {
                  "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                  "name": "Other",
                  "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
                },
              ],
              "default_category_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
              "operand": "@input.text",
              "result_name": "",
              "type": "switch",
              "wait": {
                "type": "msg",
              },
            },
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
          "ui": {
            "config": {
              "cases": {
                "b1f332f3-bdd3-4891-aec5-1843a712dbf1": {
                  "arguments": [
                    "not blue",
                  ],
                },
              },
            },
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "automatic_classify",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <div
    className="_input_99e97f"
  >
    <div
      className="_label_99e97f u font secondary body-md color-neutral-cloudy"
    >
      Enter an expression to use as the input to your classifier. To classify the last response from the contact use
      <span
        className="_link_99e97f color-weni-600"
      >
        @input.text
      </span>
    </div>
    <TextInputElement
      autocomplete={true}
      entry={
        {
          "value": "@input.text",
        }
      }
      name="Operand"
      onChange={[Function]}
      placeholder="@input.text"
      showLabel={false}
    />
  </div>
  <DescriptionAlert
    openDescriptionEdit={[Function]}
  />
  <div
    className="_content_99e97f"
  >
    <div
      className="_phrases_99e97f"
    >
      <div
        className="_header_99e97f"
      >
        <span
          className="_title_99e97f"
        >
          Command phrases
        </span>
        <span
          className="_description_99e97f"
        >
          Write command phrases related to the category
        </span>
      </div>
    </div>
    <div
      className="_categories_99e97f"
    >
      <div
        className="_header_99e97f"
      >
        <span
          className="_title_99e97f"
        >
          Categorize as
        </span>
        <span
          className="_description_99e97f"
        >
          Define the category you want to classify the contact's response
        </span>
      </div>
    </div>
  </div>
  <CaseList
    cases={
      [
        {
          "categoryName": "Red",
          "kase": {
            "arguments": [
              "red",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_any_word",
            "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          },
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          "valid": true,
        },
        {
          "categoryName": "Red",
          "kase": {
            "arguments": [
              "not blue",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_any_word",
            "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
          },
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
    required={true}
    type="smart"
  />
  <OptionalTextInput
    maxLength={64}
    name="Save as result"
    onChange={[Function]}
    toggleText="Save as"
    value={
      {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`AutomaticClassifyRouterForm > should render 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="wait_for_response"
  title="Wait for response"
>
  <TypeList
    __className=""
    initialType={
      {
        "aliases": [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "categories",
          "cases",
        ],
        "name": "Wait for response",
        "type": "wait_for_response",
        "visibility": [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    nodeSettings={
      {
        "originalAction": null,
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [],
            "exits": [
              {
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
              {
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
            ],
            "router": {
              "cases": [
                {
                  "arguments": [
                    "red",
                  ],
                  "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                  "type": "has_any_word",
                  "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
                },
              ],
              "categories": [
                {
                  "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                  "name": "Red",
                  "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                },
                {
                  "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "name": "Other",
                  "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                },
              ],
              "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              "operand": "@input.text",
              "result_name": "",
              "type": "switch",
              "wait": {
                "type": "msg",
              },
            },
            "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
          },
          "ui": {
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "wait_for_response",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <div
    className="_input_99e97f"
  >
    <div
      className="_label_99e97f u font secondary body-md color-neutral-cloudy"
    >
      Enter an expression to use as the input to your classifier. To classify the last response from the contact use
      <span
        className="_link_99e97f color-weni-600"
      >
        @input.text
      </span>
    </div>
    <TextInputElement
      autocomplete={true}
      entry={
        {
          "value": "@input.text",
        }
      }
      name="Operand"
      onChange={[Function]}
      placeholder="@input.text"
      showLabel={false}
    />
  </div>
  <DescriptionAlert
    openDescriptionEdit={[Function]}
  />
  <div
    className="_content_99e97f"
  >
    <div
      className="_phrases_99e97f"
    >
      <div
        className="_header_99e97f"
      >
        <span
          className="_title_99e97f"
        >
          Command phrases
        </span>
        <span
          className="_description_99e97f"
        >
          Write command phrases related to the category
        </span>
      </div>
    </div>
    <div
      className="_categories_99e97f"
    >
      <div
        className="_header_99e97f"
      >
        <span
          className="_title_99e97f"
        >
          Categorize as
        </span>
        <span
          className="_description_99e97f"
        >
          Define the category you want to classify the contact's response
        </span>
      </div>
    </div>
  </div>
  <CaseList
    cases={[]}
    data-spec="cases"
    onCasesUpdated={[Function]}
    required={true}
    type="smart"
  />
  <OptionalTextInput
    maxLength={64}
    name="Save as result"
    onChange={[Function]}
    toggleText="Save as"
    value={
      {
        "value": "Result",
      }
    }
  />
</Dialog>
`;

exports[`AutomaticClassifyRouterForm > updates > should not save changes if case is invalid 1`] = `
{
  "cases": [
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "red",
        ],
        "type": "has_any_word",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "valid": true,
    },
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "maroon",
        ],
        "type": "has_any_word",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
      "valid": false,
    },
    {
      "categoryName": "Green",
      "kase": {
        "arguments": [
          "green",
        ],
        "type": "has_any_word",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
      "valid": true,
    },
  ],
  "hiddenCases": [],
  "operand": {
    "validationFailures": [],
    "value": "@results.result1",
  },
  "resultName": {
    "validationFailures": [],
    "value": "Favorite Color",
  },
  "valid": false,
}
`;

exports[`AutomaticClassifyRouterForm > updates > should save changes 1`] = `
{
  "cases": [
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "red",
        ],
        "type": "has_any_word",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "valid": true,
    },
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "maroon",
        ],
        "type": "has_any_word",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
      "valid": true,
    },
    {
      "categoryName": "Green",
      "kase": {
        "arguments": [
          "green",
        ],
        "type": "has_any_word",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
      "valid": true,
    },
  ],
  "hiddenCases": [],
  "operand": {
    "validationFailures": [],
    "value": "@results.result1",
  },
  "resultName": {
    "validationFailures": [],
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`AutomaticClassifyRouterForm > updates > should save changes 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [],
          "exits": [
            {
              "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            },
            {
              "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            },
            {
              "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            },
            {
              "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "red",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "maroon",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "green",
                ],
                "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "type": "has_any_word",
              },
              {
                "arguments": [],
                "category_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
                "type": "has_category",
                "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
              },
            ],
            "categories": [
              {
                "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "name": "Red",
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
              {
                "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "name": "Green",
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
              {
                "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "name": "Failure",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
              {
                "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                "name": "Other",
                "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
              },
            ],
            "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "operand": "@results.result1",
            "result_name": "Favorite Color",
            "type": "smart",
          },
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
        "ui": {
          "config": {
            "cases": {},
          },
          "position": null,
          "type": "wait_for_response",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
